File: idl/test/preserve/driver.cpp
#include "generated/cpp/rec.h"
#include <fstream>
#include <iostream>

std::string read(const std::string& path) {
    std::ifstream f(path, std::ios::binary);
    if(!f) throw std::runtime_error("missing input");
    return std::string(std::istreambuf_iterator<char>(f), {});
}
int main(int argc,char**argv) {
    if(argc<4)return 2;
    try {
        auto r=rec::decode(read(argv[1]));const std::string scope=argv[3];
        if(scope=="mutate") {r.id="edited";if(r.child)r.child->name="edited-child";for(auto& child:r.children)child.name="edited-child";}
        else if(scope!="read") {
            auto* target=&r.extras;
            if(scope=="child")target=&r.child->extras;else if(scope=="repeated")target=&r.children[0].extras;
            const std::string key=scope=="bad-key"?std::string(1,char(255)):argv[4];
            (*target)[key]=read(argv[5]);
        }
        auto output=rec::encode(r);rec::decode(output);
        std::ofstream f(argv[2],std::ios::binary);f<<output;
        if(!f)throw std::runtime_error("output failed");
        std::cout<<"ok";
    }catch(const rec::Refusal&e){std::cout<<e.word;}
}

File: idl/test/preserve/driver.go
package main

import (
	"fmt"
	"os"
	rec "preserve.test/generated/go/rec"
)

func main() {
	defer func() {
		if e := recover(); e != nil {
			if r, ok := e.(*rec.Refusal); ok {
				fmt.Print(r.Word)
				return
			}
			panic(e)
		}
	}()
	input, err := os.ReadFile(os.Args[1])
	if err != nil {
		panic(err)
	}
	r, err := rec.Decode(input)
	if err != nil {
		fmt.Print(err.(*rec.Refusal).Word)
		return
	}
	scope := os.Args[3]
	if scope == "mutate" {
		r.Id = "edited"
		if r.Child != nil {
			r.Child.Name = "edited-child"
		}
		for i := range r.Children {
			r.Children[i].Name = "edited-child"
		}
	}
	if scope != "mutate" && scope != "read" {
		raw, err := os.ReadFile(os.Args[5])
		if err != nil {
			panic(err)
		}
		target := &r.Extras
		if scope == "child" {
			target = &r.Child.Extras
		}
		if scope == "repeated" {
			target = &r.Children[0].Extras
		}
		if *target == nil {
			*target = map[string]rec.Raw{}
		}
		key := os.Args[4]
		if scope == "bad-key" {
			key = string([]byte{255})
		}
		(*target)[key] = rec.Raw(raw)
	}
	output := rec.Encode(r)
	if _, err := rec.Decode(output); err != nil {
		panic(err)
	}
	if err := os.WriteFile(os.Args[2], output, 0600); err != nil {
		panic(err)
	}
	fmt.Print("ok")
}
